### PR TITLE
MAGN-10353 Crash when installing packages to folder (Network Path Issue)

### PR DESCRIPTION
**Pull request by NightHawk-Michael**
### Purpose
- Solve the problem of inaccessible network path causing hard crash on Dynamo by catching the exception.
- Since the inaccessible network path is still default user directory in the setting, the attempt to save a newly created custom node will show an error dialog indicating that the path is not accessible. However, the dialog is unwanted in this case. Hence, we have decided to automatically choose the next priority path in the Manage Node and Package Path and set it to be the initial path of the file save dialog. 
- In the case of there is only one path in Manage Node and Package Path and that path is not accessible, the initial path of the file save dialog will be anywhere that user recently visit. This behavior is handled by Windows.
- FYI: Here is the old pull request https://github.com/DynamoDS/Dynamo/pull/6973 that fixed the same problem in local directory but not network path. 

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Follows [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) standards (*No change of public methods or types etc..*)

### Reviewers
@ke-yu 

### FYIs

@Benglin

### CODE REVIEW

File: src/DynamoUtilities/PathHelper.cs [hunk 1]
```diff
@@ -13,8 +13,13 @@ public static Exception CreateFolderIfNotExist(string folderPath)
         {
             try
             {
+                // When network path is access denied, the Directory.Exits however still 
+                // return true.
+                // EnumerateDirectories operation is additional check
+                // to catch exception for network path.
                 if (!Directory.Exists(folderPath))
                     Directory.CreateDirectory(folderPath);
+                Directory.EnumerateDirectories(folderPath);
```
Comment (ke-yu) [inline src/DynamoUtilities/PathHelper.cs]: Why call this function without using its return value?

File: src/DynamoPackages/PackageLoader.cs [hunk 2]
```diff
@@ -203,19 +203,23 @@ private void ScanAllPackageDirectories(IPreferences preferences)
...
-                Directory.EnumerateDirectories(root, "*", SearchOption.TopDirectoryOnly))
-            {
-                var pkg = ScanPackageDirectory(dir);
-                if (pkg != null && preferences.PackageDirectoriesToUninstall.Contains(dir)) 
-                    pkg.MarkForUninstall(preferences);
+                foreach (var dir in
+                    Directory.EnumerateDirectories(root, "*", SearchOption.TopDirectoryOnly))
+                {
+                    var pkg = ScanPackageDirectory(dir);
+                    if (pkg != null && preferences.PackageDirectoriesToUninstall.Contains(dir))
```
Comment (ke-yu) [inline src/DynamoPackages/PackageLoader.cs]: Would `ScanPackageDirectory()` be an expensive operation? If it is, you may check `preferences.PackageDirectoriesToUninstall.Contains(dir)` before `ScanPackageDirectory()`

File: src/DynamoPackages/PackageLoader.cs [hunk 3]
```diff
@@ -203,19 +203,23 @@ private void ScanAllPackageDirectories(IPreferences preferences)
...
+                        pkg.MarkForUninstall(preferences);
+                }
             }
+            catch (Exception ex) { }
```
Comment (ke-yu) [inline src/DynamoPackages/PackageLoader.cs]: Catch exception and do nothing?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 4]
```diff
@@ -1592,7 +1592,24 @@ public void ShowSaveDialogAndSaveResult(object parameter)
             else if (vm.Model.CurrentWorkspace is CustomNodeWorkspaceModel)
             {
                 var pathManager = vm.model.PathManager;
-                _fileDialog.InitialDirectory = pathManager.DefaultUserDefinitions;
+                try
+                {
+                    Directory.EnumerateDirectories(pathManager.DefaultUserDefinitions);
```
Comment (ke-yu) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Return value not used?

File: src/DynamoUtilities/PathHelper.cs (hunk 1 above)
Comment (gregmarr) [inline src/DynamoUtilities/PathHelper.cs]: Per the comment, I assume it throws an exception if there is an access problem.

**Pull request merged**